src/dwarf/parser.rs

```diff
@@ -684,136 +700,186 @@ fn find_die_sibling(die: &mut debug_info::DIE<'_>) -> Option<usize> {
 /// * `str_data` - is the content of the `.debug_str` section.
 ///
 /// Return a [`DWSymInfo`] if it finds the address of the subprogram.
-fn parse_die_subprogram<'a>(
-    die: &mut debug_info::DIE<'a>,
-    str_data: &'a [u8],
-) -> Result<Option<DWSymInfo<'a>>> {
-    let mut addr: Option<Addr> = None;
-    let mut name_str: Option<&str> = None;
-    let mut size = 0;
-
-    for (name, _form, _opt, value) in die {
-        match name {
+// TODO: Having a single function for a single subprogram may not be
+//       sufficient to get all relevant symbol information. See
+//       https://stackoverflow.com/a/59674438
+// TODO: We likely need to handle DW_AT_ranges; see
+//       https://reviews.llvm.org/D78489
+fn parse_die_subprogram<'dat>(
+    entry: &DebuggingInformationEntry<R<'dat>>,
+    debug_str: &DebugStr<R<'dat>>,
+) -> Result<Option<DWSymInfo<'dat>>> {
+    let mut addr = None;
+    let mut name = None;
+    let mut size = None;
+    let mut high_pc = None;
+    let mut linkage_name = None;
+
+    let mut attrs = entry.attrs();
+    while let Some(attr) = attrs.next().map_err(|err| {
+        Error::new(
+            ErrorKind::InvalidData,
+            format!("failed to read next DIE attribute: {err}"),
+        )
+    })? {
+        match attr.name() {
             constants::DW_AT_linkage_name | constants::DW_AT_name => {
-                if name_str.is_some() {
-                    continue
-                }
-                name_str = Some(match value {
-                    debug_info::AttrValue::Unsigned(str_off) => unsafe {
-                        CStr::from_ptr(str_data[str_off as usize..].as_ptr().cast())
-                            .to_str()
-                            .map_err(|_e| {
-                                Error::new(
-                                    ErrorKind::InvalidData,
-                                    "fail to extract the name of a subprogram",
-                                )
-                            })?
-                    },
-                    debug_info::AttrValue::String(s) => s,
+                let attr_name = attr
```

Comment:
we seem to need this only for error/warning reporting, yet we pay the overhead each time. Can this be done lazily?

Reply:
reading further, seems like exactly what you are doing for DW_AT_low_pc and DW_AT_high_pc, so why not the same here?

Reply:
I just didn't want to repeat the same thing three times. But will move that calculation into a function.